patroni/postgresql.py

```diff
@@ -759,27 +789,111 @@ def checkpoint(self, connect_kwargs=None):
+
+        # We can skip safepoint detection if nobody is waiting for it.
+        if not self.stop_safepoint_reached.is_set():
+            # Wait for our connection to terminate so we can be sure that no new connections are being initiated
+            self._wait_for_connection_close(pid)
+            self._wait_for_user_backends_to_close(pid)
+            self.stop_safepoint_reached.set()
+
+        self._wait_for_postmaster_stop(pid)
+
+        return True, True
+
+    def _wait_for_postmaster_stop(self, pid):
+        # This wait loop differs subtly from pg_ctl as we check for both the pid file going
+        # away and if the pid is running. This seems safer.
+        while pid == self.get_pid() and self.is_pid_running(pid):
+            time.sleep(STOP_POLLING_INTERVAL)
+
+    def _signal_postmaster_stop(self, mode):
+        pid = self.get_pid()
+        if pid == 0:
+            return None, True
+        elif pid < 0:
+            logger.warning("Cannot stop server; single-user server is running (PID: {0})".format(-pid))
+            return None, False
+        try:
+            os.kill(pid, STOP_SIGNALS[mode])
+        except OSError as e:
+            if e.errno == errno.ESRCH:
+                return None, True
+            else:
+                logger.warning("Could not send stop signal to PostgreSQL (error: {0})".format(e.errno))
+                return None, False
+        return pid, None
+
+    def terminate_starting_postmaster(self, pid):
+        """Terminates a postmaster that has not yet opened ports or possibly even written a pid file. Blocks
+        until the process goes away."""
+        try:
+            os.kill(pid, STOP_SIGNALS['immediate'])
+        except OSError as e:
+            if e.errno == errno.ESRCH:
+                return
+            logger.warning("Could not send stop signal to PostgreSQL (error: {0})".format(e.errno))
+
+        while self.is_pid_running(pid):
+            time.sleep(STOP_POLLING_INTERVAL)
+
+    def _wait_for_connection_close(self, pid):
+        try:
+            with self.connection().cursor() as cur:
+                while True:  # Need a timeout here?
+                    if pid == self.get_pid() and self.is_pid_running(pid):
+                        cur.execute("SELECT 1")
+                        time.sleep(STOP_POLLING_INTERVAL)
+                        continue
+                    else:
+                        break
+        except psycopg2.Error:
+            pass
+
+    @staticmethod
+    def _wait_for_user_backends_to_close(postmaster_pid):
+        # These regexps are cross checked against versions PostgreSQL 9.1 .. 9.6
+        aux_proc_re = re.compile("(?:postgres:)( .*:)? (?:""(?:startup|logger|checkpointer|writer|wal writer|"
+                                 "autovacuum launcher|autovacuum worker|stats collector|wal receiver|archiver|"
+                                 "wal sender) process|bgworker: )")
+
+        postmaster = psutil.Process(postmaster_pid)
```

Comment:
> Thanks, that was indeed broken, but not the cause of test failures. 

Actually it was. According to watchdog logs Patroni shut down was taking ~120 seconds and "magic close" wasn't received from the pipe.

120 seconds comes from 60 seconds * 2, where 2 is a default timeout multiplier when we run tests on travis (sometimes travis is very slow).
features/environment.py sends a SIGTERM to Patroni and waits for 120 seconds until it will shut down normally. After timeout it sends a SIGKILL.

Patroni can't send "magic close" if psutil.Process raise an exception. And Patroni also doesn't shut down normally in this case, because one or more of the threads didn't terminated.
```
features/output/watchdog/watchdog.postgres0.log:2017-04-12 09:57:08,557218: watchdog postgres0 initialized
features/output/watchdog/watchdog.postgres0.log:2017-04-12 09:57:08,574592: starting fifo /home/travis/build/zalando/patroni/data/watchdog.postgres0.fifo
features/output/watchdog/watchdog.postgres0.log:2017-04-12 09:57:08,575271: opening
features/output/watchdog/watchdog.postgres0.log:2017-04-12 09:57:08,897032: Fifo /home/travis/build/zalando/patroni/data/watchdog.postgres0.fifo connected
features/output/watchdog/watchdog.postgres0.log:2017-04-12 09:57:08,897244: timeout=15
features/output/watchdog/watchdog.postgres0.log:2017-04-12 09:57:08,901129: ping after 0.326314926147 seconds
features/output/watchdog/watchdog.postgres0.log:2017-04-12 09:59:11,742916: Pipe closed
features/output/watchdog/watchdog.postgres0.log:2017-04-12 09:59:11,743001: closing
features/output/watchdog/watchdog.postgres0.log:2017-04-12 09:59:11,743034: opening
features/output/watchdog/watchdog.postgres0.log:2017-04-12 09:59:12,740208: Monitor stop
```